kroto-plus-coroutines/src/main/kotlin/com/github/marcoferrer/krotoplus/coroutines/call/FlowControl.kt

```diff
@@ -42,40 +43,48 @@ internal fun <T> CallStreamObserver<*>.applyInboundFlowControl(
     }
 }
 
+private typealias MessageHandler = suspend ActorScope<*>.() -> Unit
+
 internal fun <T> CoroutineScope.applyOutboundFlowControl(
     streamObserver: CallStreamObserver<T>,
     targetChannel: Channel<T>
 ){
-    val isOutboundJobRunning = AtomicBoolean()
+
+    val isCompleted = AtomicBoolean()
     val channelIterator = targetChannel.iterator()
-    streamObserver.setOnReadyHandler {
-        if(targetChannel.isClosedForReceive){
-            streamObserver.completeSafely()
-        }else if(
+    val messageHandlerBlock: MessageHandler = handler@ {
+        while(
             streamObserver.isReady &&
-            !targetChannel.isClosedForReceive &&
-            isOutboundJobRunning.compareAndSet(false, true)
+            channelIterator.hasNext()
         ){
-            launch(Dispatchers.Unconfined + CoroutineExceptionHandler { _, e ->
-                streamObserver.completeSafely(e)
-                targetChannel.close(e)
-            }) {
-                try{
-                    while(
-                        streamObserver.isReady &&
-                        !targetChannel.isClosedForReceive &&
-                        channelIterator.hasNext()
-                    ){
-                        val value = channelIterator.next()
-                        streamObserver.onNext(value)
-                    }
-                    if(targetChannel.isClosedForReceive){
-                        streamObserver.onCompleted()
-                    }
-                } finally {
-                    isOutboundJobRunning.set(false)
-                }
-            }
+            streamObserver.onNext(channelIterator.next())
+        }
+        if(targetChannel.isClosedForReceive && isCompleted.compareAndSet(false,true)){
+            streamObserver.onCompleted()
+            channel.close()
+        }
+    }
+
+    val messageHandlerActor = actor<MessageHandler>(
+        capacity = Channel.UNLIMITED,
+        context = Dispatchers.Unconfined + CoroutineExceptionHandler { _, e ->
+            streamObserver.completeSafely(e)
+            targetChannel.close(e)
+        }
+    ) {
+        for(handler in channel){
+            if(isCompleted.get()) break
+            handler(this)
+        }
```

Comment:
Adding this here reduced the problem but didn't eliminate it. I guess there must be other paths for the scope to cancel.
```
       catch(ex: CancellationException) {
            targetChannel.cancel(ex)
            throw ex
        }
```

Reply:
So the scope could cancellation can also be propagated from its parent under normal normal coroutine usage. This case was covered before because executing new launch on a cancelled scope would take care. Its hard to reproduce but Im trying a few things now.